File: c++/15787.cpp
#include <iostream>
#include <set>

using namespace std;

int train[100001];

set<int> s;

int main() {
  freopen("example.txt", "r", stdin);
  int N, M, i, x;
  cin >> N >> M;

  while (M--) {
    int c;
    cin >> c;

    if (c == 1) {
      cin >> i >> x;
      train[i] |= (1 << x);
    } else if (c == 2) {
      cin >> i >> x;
      train[i] &= ~(1 << x);
    } else if (c == 3) {
      cin >> i;
      train[i] <<= 1;
      train[i] &= ~(1 << 21);
    } else if (c == 4) {
      cin >> i;
      train[i] >>= 1;
      train[i] &= ~1;
    }
  }

  for (int i = 1; i <= N; i++) {
    s.insert(train[i]);
  }

  cout << s.size();

  return 0;
}

File: js/15787.js
const fs = require('fs');

const inputs = fs
  .readFileSync('example.txt')
  .toString()
  .trim()
  .split('\n')
  .map((e) => e.split(' ').map(Number));

let idx = 0;
const getLine = () => inputs[idx++];
const s = new Set();

let [N, M] = getLine();

const train = Array(N + 1).fill(0);

while (M--) {
  const line = getLine();

  if (line[0] === 1) {
    train[line[1]] |= 1 << line[2];
  } else if (line[0] === 2) {
    train[line[1]] &= ~(1 << line[2]);
  } else if (line[0] === 3) {
    train[line[1]] <<= 1;
    train[line[1]] &= ~(1 << 21);
  } else if (line[0] === 4) {
    train[line[1]] >>= 1;
    train[line[1]] &= ~1;
  }
}

for (let i = 1; i <= N; i++) {
  s.add(train[i]);
}

console.log(s.size);
